File: src/u_ocaml-output/FStar_Format.ml

open Prims
type doc =
| Doc of Prims.string


let uu___is_Doc : doc  ->  Prims.bool = (fun projectee -> true)


let __proj__Doc__item___0 : doc  ->  Prims.string = (fun projectee -> (match (projectee) with
| Doc (_0) -> begin
_0
end))


let empty : doc = Doc ("")


let hardline : doc = Doc ("\n")


let text : Prims.string  ->  doc = (fun s -> Doc (s))


let num : Prims.int  ->  doc = (fun i -> Doc ((Prims.string_of_int i)))


let break_ : Prims.int  ->  doc = (fun i -> Doc (""))


let break0 : doc = (break_ (Prims.parse_int "0"))


let break1 : doc = (text " ")


let enclose : doc  ->  doc  ->  doc  ->  doc = (fun uu____26 uu____27 uu____28 -> (match (((uu____26), (uu____27), (uu____28))) with
| (Doc (l), Doc (r), Doc (x)) -> begin
Doc ((Prims.strcat l (Prims.strcat x r)))
end))


let brackets : doc  ->  doc = (fun uu____34 -> (match (uu____34) with
| Doc (d) -> begin
(enclose (text "[") (text "]") (Doc (d)))
end))


let cbrackets : doc  ->  doc = (fun uu____38 -> (match (uu____38) with
| Doc (d) -> begin
(enclose (text "{") (text "}") (Doc (d)))
end))


let parens : doc  ->  doc = (fun uu____42 -> (match (uu____42) with
| Doc (d) -> begin
(enclose (text "(") (text ")") (Doc (d)))
end))


let cat : doc  ->  doc  ->  doc = (fun uu____48 uu____49 -> (match (((uu____48), (uu____49))) with
| (Doc (d1), Doc (d2)) -> begin
Doc ((Prims.strcat d1 d2))
end))


let reduce : doc Prims.list  ->  doc = (fun docs -> (FStar_List.fold_left cat empty docs))


let group : doc  ->  doc = (fun uu____59 -> (match (uu____59) with
| Doc (d) -> begin
Doc (d)
end))


let groups : doc Prims.list  ->  doc = (fun docs -> (group (reduce docs)))


let combine : doc  ->  doc Prims.list  ->  doc = (fun uu____71 docs -> (match (uu____71) with
| Doc (sep) -> begin
(

let select = (fun uu____79 -> (match (uu____79) with
| Doc (d) -> begin
(match ((d = "")) with
| true -> begin
None
end
| uu____83 -> begin
Some (d)
end)
end))
in (

let docs = (FStar_List.choose select docs)
in Doc ((FStar_String.concat sep docs))))
end))


let cat1 : doc  ->  doc  ->  doc = (fun d1 d2 -> (reduce ((d1)::(break1)::(d2)::[])))


let reduce1 : doc Prims.list  ->  doc = (fun docs -> (combine break1 docs))


let nest : Prims.int  ->  doc  ->  doc = (fun i uu____102 -> (match (uu____102) with
| Doc (d) -> begin
Doc (d)
end))


let align : doc Prims.list  ->  doc = (fun docs -> (

let uu____109 = (combine hardline docs)
in (match (uu____109) with
| Doc (doc) -> begin
Doc (doc)
end)))


let hbox : doc  ->  doc = (fun d -> d)


let pretty : Prims.int  ->  doc  ->  Prims.string = (fun sz uu____119 -> (match (uu____119) with
| Doc (doc) -> begin
doc
end))

File: src/format/FStar.Format.fs
(* -------------------------------------------------------------------- *)
#light "off"

module FStar.Format
open FStar.ST
open FStar.All
open FStar
open FStar.String

(* -------------------------------------------------------------------- *)
type doc = | Doc of string



(* -------------------------------------------------------------------- *)
let empty    = Doc ""
let hardline = Doc "\n"

(* -------------------------------------------------------------------- *)
let text (s : string) = Doc s
let num (i : int) = Doc (string_of_int i)

(* -------------------------------------------------------------------- *)
let break_ (i : int   ) = Doc ""

let break0 = break_ 0
let break1 = text " "

(* -------------------------------------------------------------------- *)
let enclose (Doc l) (Doc r) (Doc x) =
    Doc (l^x^r)

let brackets (Doc d ) = enclose (text "[") (text "]") (Doc d)
let cbrackets (Doc d) = enclose (text "{") (text "}") (Doc d)
let parens   (Doc d ) = enclose (text "(") (text ")") (Doc d)

(* -------------------------------------------------------------------- *)
let cat (Doc d1) (Doc d2) = Doc (d1 ^ d2)

(* -------------------------------------------------------------------- *)
let reduce (docs : list<doc>) =
  List.fold_left cat empty docs

(* -------------------------------------------------------------------- *)
let group (Doc d) = Doc (d)

(* -------------------------------------------------------------------- *)
let groups (docs : list<doc>) =
  group (reduce docs)

(* -------------------------------------------------------------------- *)
let combine (Doc sep) (docs : list<doc>) =
  let select (Doc d) = if d = "" then None else Some d in
  let docs = List.choose select docs in
  Doc (String.concat sep docs)

(* -------------------------------------------------------------------- *)
let cat1 (d1 : doc) (d2 : doc) =
    reduce [d1; break1; d2]

(* -------------------------------------------------------------------- *)
let reduce1 (docs : list<doc>) =
    combine break1 docs

(* -------------------------------------------------------------------- *)
let nest (i : int) (Doc d) =
    Doc (d)

(* -------------------------------------------------------------------- *)
let align (docs : list<doc>) =
    let (Doc doc) = combine hardline docs in
    Doc (doc)

(* -------------------------------------------------------------------- *)
let hbox (d : doc) = d (* FIXME *)

(* -------------------------------------------------------------------- *)
let pretty (sz : int) (Doc doc) = doc
